### PR TITLE
Spess grid

### LINKED ISSUE #5614: Space transitions are not randomized
When you change levels you always transition to the same axis, creating a loop. This is terrible.

### PR DESCRIPTION
**Pull request by norill**
Arranged all playable z levels (SS13, derelict, old sat and mining) in a grid. Removed z6 and z7.
The grid is normally 2x2, but if the number of z levels is other than 4 it might be 1x2, 2x1, 2x3 or 3x2. 
It's randomized, but all transitions are bidirectional. Fixes #5614

My first real coding in DM, don't merge without testing.

### CODE REVIEW

**Comment by Jordie0608** [comment]
Removing z6 and z7 will break a bunch of things, namely space transitions and falling off the shuttle.

**Comment by WJohn** [comment]
Easier to understand than the spacecube and has bidirectional movement :+1:

**Comment by norill** [comment]
>Removing z6 and z7 will break a bunch of things, namely space transitions and falling off the shuttle.


this pr is all about space transitions, and i got falling off the shuttle covered too:
>- falling off shuttle in transit will move you to random accessible level (including SS13 now)

**Comment by pudl** [comment]
thank you based stranger.

**Comment by Incoming5643** [comment]
<b>YOU CAN'T CUCK THE CUBE, NOT THE CUUUUUUUUUBE</b>

Just kidding that's fine I guess, but I don't think you should cut the empty levels, 2x2 is a bit TOO navigable if you ask me, a four unit square of z levels only has a few meaningful arrangements and very little distance between points.

From standing on any one z level there's only three positions in terms of layouts: The level that lies to the north and south, The level that lies to the east and west, and the level in the corner you can't reach by traveling in a straight line alone.

From station:
Mining east/west | Derelict north/south | Old AI sat Corner
Derelict east/west | Mining north/south | Old AI sat Corner
Mining east/west | Old AI sat north/south | Derelict Corner
Old AI sat east/west | Mining north/south | Derelict Corner
Derelict east/west | Old AI sat north/south | Mining Corner
Old AI sat east/west | Derelict north/south | Mining Corner

While there's other layouts based on where the station is on the foursquare, they aren't actually perceived as a meaningful difference from the point of view of a spessmen.

Keeping the two empty levels helps resolve this predictability of space a little bit, but then space becomes an 3x2 (or 1x5?) obloid. The hitler in me says you should make it 3x3 by having a grand total of 5 empty space z levels to get lost in and then really open the floor to adding more mapped shit to empty space buuuuuuuuut I'd understand people's resistance to this.

**Comment by Miauw62** [comment]
I agree with keeping the two empty levels in.
Op 7-apr.-2015 18:21 schreef "Incoming5643" <notifications@github.com>:

**Comment by Iamgoofball** [comment]
While removing the 2 empty levels will speed up compile times, I'd prefer if we put something there instead of removing them.

**Comment by ZomgPonies** [comment]
Ideally since this should make the whole of space much more navigable, you should probably be looking at a 3x3 grid.

**Comment by RemieRichards** [comment]
so THIS is how you found the bias in our ```shuffle()```
I like the idea that we get navigable random space, as opposed to the broken cube, :+1: 

I do think expanding out to a 3x3 of blank space levels (which we can fill with junk/crap) is better, as has been mentioned previously.

File: code/game/turfs/space/space.dm [hunk 1]
```diff
@@ -8,16 +8,14 @@
...
 
-	var/transition //These is used in transistions as a way to tell where on the "cube" of space you're transitioning from/to
 	var/destination_x
 	var/destination_y
+	var/destination_z
 
 /turf/space/New()
 	if(!istype(src, /turf/space/transit))
 		icon_state = "[((x + y) ^ ~(x * y) + z) % 25]"
-	if(config)
```
Comment (RemieRichards) [inline code/game/turfs/space/space.dm]: Why have you removed the config check for starlight?

Comment (norill) [inline code/game/turfs/space/space.dm]: it was superfluous, because there is another one in update_starlight

Comment (RemieRichards) [inline code/game/turfs/space/space.dm]: Fair enough :+1:

**Comment by WJohn** [comment]
I'd say we first fill up our other z levels before keeping the two extra empty ones. Our current z levels are already damn near empty, boasting at most 2 or 3 small things on them.

**Comment by RemieRichards** [comment]
@Wjohn our current Z levels are fine.
They have *enough* content, It's better for a z level to have Some content Vs. Too much.

**Comment by norill** [comment]
re: ez navigation - keep in mind that:
- drifting all the way through a level takes about 2 minutes of staring at the screen while absolutely nothing interesting happens
- old sat and derelict are mostly space
- you can only see 7 tiles in each direction
- there is no way to tell which z level you are on

all above makes deep space travel synonymous to cancer

**Comment by Lumipharon** [comment]
gps's will tell you what z-level you're on.

We need empty z-levels otherwise space is essentially crowded and easy to navigate.

**Comment by Incoming5643** [comment]
if you ask me space should be hopelessly huge and mostly empty, because that's how real space is. The new z levels should mostly contain small islands of interest in the huge void.

Also there's some wacky shit that can happen if the "universe" is small enough like people throwing into space reappearing on the other side before they die and singuloth looping around for multiple passes at the station (this already happens occasionally, and is hilarious)

**Comment by optimumtact** [comment]
DEATH TO Z6 and Z7

THEY MAKE THE GAME SHIT

**Comment by Incoming5643** [comment]
"you know what I hate about space? It's too big."

**Comment by optimumtact** [comment]
"you know what I hate about big empty spaces? It's not a fun game mechanic"

**Comment by RemieRichards** [comment]
"You know what I hate about content? Too much of it in one place at one time"

**Comment by Lumipharon** [comment]
having everything in space easily reachable makes it just that - easily and reliabl reachable, which makes it less fun.

Nothing in space is essential - if you're going into space to look for stuff you're making the active choice to leave station to do so.

Being able to get lose, or drift for ever is good.
Having bodies land back at the station every time is bad.

**Comment by norill** [comment]
adding empty space won't affect the chance of bodies getting back to the station. also there are so many alternatives to spacing bodies

**Comment by Lumipharon** [comment]
I mean compared to our current system.
With the current, random on round start z-level edges, if you throw a body off one edge, there is a good chance that it will not come back to the station z-level.

If it's a fixed grid, assuming you throw it in a straight line (cardinal directions) unless it hits something else first, it WILL come back to the station.

**Comment by norill** [comment]
our current system is broken

**Comment by Lumipharon** [comment]
yes, but that doesn't change my point.

**Comment by Incoming5643** [comment]
It's not actually broken, I just intentionally didn't fix it to be bidirectional because that would force the spessman to in many cases suddenly change their perceived direction while transferring z-levels, which is confusing in its own right.

Before SPACE CUBE there was RANDOM SPACE, where you just had a prob chance to end up on any of the potential z-levels every time you transitioned. It was crappy and utterly unpredictable.

But before RANDOM SPACE there was basically THIS system. The essential z levels sat in a foursquare and while they weren't shuffled then it still lead to the same sort of small world problem I intentionally tried to avoid with SPACE CUBE.

The reason RANDOM SPACE happened was because people hated how small and well mapped space was. It sat out there like a big galactic maintenance run of goodies and there was this whole inane "WE HAVE TO MAKE THEM HATE GOING TO SPACE" movement. Eventually people grew to hate RANDOM SPACE it because RNG would tend to fuck people over and say "ok now transisition through z = 6 five times". I still think people would prefer BIG SPACE though maybe I'm wrong? It's hard to tell.

**Comment by pudl** [comment]
im just gonna say if space becomes a 60% empty 3x3 grid i will kill myself on stream

**Comment by Miauw62** [comment]
The thing with bodies is a problem yeah. I suggest we keep the two empty
levels and add some more space content Soon (TM), so the chances of bodies
getting stuck are increased.
Op 8-apr.-2015 15:44 schreef "Annie Acosta" <notifications@github.com>:

**Comment by Jalleo** [comment]
How about we do this we have 3x3 space but one z missing like a puzzle every so often a z levl will move into that empty space changing where everything can end up. This makes it semi reliable while not having it too random or confusing.

**Comment by Incoming5643** [comment]
SLIDE PUZZLE SPACE?

I hate this silliness so much less than I should!

**Comment by Jalleo** [comment]
Its because it actually makes sense and probably could have multiple blank spaces to make travel possibly more random. (could cause people getting stuck if the station was at center and other 4 zs were in the corner. If there were no shuttles and teleporters that is.

**Comment by Metacide** [comment]
Please ping me when this is done / pulled.

**Comment by Razharas** [comment]
Please no randomized space travel
Its the hitler

File: code/game/turfs/space/transit.dm [hunk 2]
```diff
@@ -11,7 +11,7 @@
 	var/max = world.maxx-TRANSITIONEDGE
 	var/min = 1+TRANSITIONEDGE
 
-	var/_z = rand(ZLEVEL_SPACEMIN,ZLEVEL_SPACEMAX)	//select a random space zlevel
+	var/_z = pick(accessible_z_levels)	//select a random space zlevel
```
Comment (MrPerson) [inline code/game/turfs/space/transit.dm]: If this picks an occupied z, it's quite possible you'll literally land in the middle of the station or worse, the asteroid. There's a reason this dumped you off into space.

Comment (norill) [inline code/game/turfs/space/transit.dm]: nah, it always dumps you on the edge

**Comment by Razharas** [comment]
Whats the status of this?

**Comment by Razharas** [comment]
Ok i make a judgement since none cares - denied

**Pull request closed without merging**